### ISSUE TITLE
[AzureStack] debug traces hard codes the arm endpoint management.azure.com

### ISSUE THREAD

**Issue by bganapa** [Azure/autorest.powershell#558]
The call is actually going to azurestack endpoint correctly, but the debug traces is still pointing to management.azure.com

PS C:\> Get-AzsBackupConfiguration -Location local -ResourceGroupName system.local
DEBUG: 10:23:23 AM - GetAzureRMContextCommand end processing.
DEBUG: 10:23:23 AM - GetAzureRMContextCommand end processing.
DEBUG: CmdletBeginProcessing: 
DEBUG: CmdletProcessRecordStart: 
DEBUG: [CmdletProcessRecordAsyncStart]: Created new QosEvent for command 'Get-AzsBackupConfiguration_Get': AzureQoSEvent: CommandName - Get-AzsBackupConfiguration_G
et; IsSuccess - False; Duration - 00:00:00; Exception - ;
DEBUG: CmdletProcessRecordAsyncStart: 
DEBUG: CmdletGetPipeline: 
DEBUG: CmdletBeforeAPICall: 
DEBUG: URLCreated: https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/back
upLocations/local?api-version=2018-09-01
DEBUG: RequestCreated: https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/
backupLocations/local?api-version=2018-09-01
DEBUG: HeaderParametersAdded: https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup
.Admin/backupLocations/local?api-version=2018-09-01
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
GET

Absolute Uri:
https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/backupLocations/local?a
pi-version=2018-09-01

Headers:

Body:

DEBUG: BeforeCall: 
DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
OK

Headers:
Pragma                        : no-cache
x-ms-correlation-request-id   : 2931b1f4-7b74-4c7c-9a3e-9347fad18d27
x-ms-ratelimit-remaining-subscription-reads: 14999
x-ms-request-id               : 2931b1f4-7b74-4c7c-9a3e-9347fad18d27
x-ms-routing-request-id       : LOCAL:20200121T182325Z:2931b1f4-7b74-4c7c-9a3e-9347fad18d27
Strict-Transport-Security     : max-age=31536000; includeSubDomains
X-Content-Type-Options        : nosniff
Cache-Control                 : no-cache
Date                          : Tue, 21 Jan 2020 18:23:25 GMT
Server                        : Microsoft-HTTPAPI/2.0
WWW-Authenticate              : oYG3MIG0oAMKAQChCwYJKoZIgvcSAQICooGfBIGcYIGZBgkqhkiG9xIBAgICAG+BiTCBhqADAgEFoQMCAQ+iejB4oAMCARKicQRvn1NXLhZPqsDvwOUVjKEgu1D0SC4Qqq41
P0FDW734LoNLQkKTzdYdnWE1v2trbmONmd34LmRtJZV9iwve0JTRXWvFZJuPI+W5sw5vmnYO0VDDZZvrkrp2XQdEbNWjPSaUxdamZ0SX62VS3tSjKT+Y

Body:
{
  "id": "/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/backupLocations/local",
  "name": "local",
  "type": "Microsoft.Backup.Admin/backupLocations",
  "location": "local",
  "tags": {},
  "properties": {
    "externalStoreDefault": {
      "path": "\\\\su1fileserver\\SU1_Infrastructure_1\\864df438-1db5-4050-be0f-7cbf570fe2a6",
      "userName": "azurestack.local\\AzureStackAdmin",
      "password": null,
      "encryptionCertBase64": null,
      "encryptionCertThumbprint": "AC07D5AF004988902A04E6D2A924ACA4D2408372",
      "backupFrequencyInHours": 4,
      "availableCapacity": "2.01 TB",
      "isBackupSchedulerEnabled": false,
      "nextBackupTime": "2020-01-21T20:20:46.8216683Z",
      "lastBackupTime": "2020-01-14T20:57:08.6116021Z",
      "backupRetentionPeriodInDays": 2
    }
  }
}

DEBUG: ResponseCreated: 
DEBUG: BeforeResponseDispatch: 

AvailableCapacity           : 2.01 TB
BackupFrequencyInHours      : 4
BackupRetentionPeriodInDays : 2
EncryptionCertBase64        : 
EncryptionCertThumbprint    : AC07D5AF004988902A04E6D2A924ACA4D2408372
Id                          : /subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/backupLocations/loca
                              l
IsBackupSchedulerEnabled    : False
LastBackupTime              : 1/14/2020 8:57:08 PM
Location                    : local
Name                        : local
NextBackupTime              : 1/21/2020 8:20:46 PM
Password                    : 
Path                        : \\su1fileserver\SU1_Infrastructure_1\864df438-1db5-4050-be0f-7cbf570fe2a6
Tag                         : Microsoft.Azure.PowerShell.Cmdlets.BackupAdmin.Models.Api20180901.ResourceTags
Type                        : Microsoft.Backup.Admin/backupLocations
UserName                    : azurestack.local\AzureStackAdmin

DEBUG: Finally: 
DEBUG: CmdletAfterAPICall: 
DEBUG: [CmdletProcessRecordAsyncEnd]: Sending new QosEvent for command 'Get-AzsBackupConfiguration_Get': AzureQoSEvent: CommandName - Get-AzsBackupConfiguration_Get
; IsSuccess - True; Duration - 00:00:00; Exception - ;
DEBUG: CmdletProcessRecordAsyncEnd: 
DEBUG: CmdletProcessRecordEnd: 
DEBUG: CmdletEndProcessing:

**Comment by bganapa** [comment]
@dolauli Can this be fixed? This totally misleads the user when they see debug traces

**Comment by bganapa** [comment]
@dolauli Can you please prioritize a fix for this issue? We are in the process of planning for our next release and we would like to get this fixed

**Comment by bganapa** [comment]
@dolauli  Can you pls update here?

**Comment by dolauli** [comment]
This root cause is clear. When printing the log, the endpoint has not been replaced by az.accounts. I would suggest we do not print the endpoint before the endpoint is replaced in the az.accounts.

**Comment by isra-fel** [comment]
We have two parts in the log that would print URLs:

1. First part shows how the request was constructed, e.g. 
```log
DEBUG: CmdletBeforeAPICall:
DEBUG: URLCreated: https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/back
upLocations/local?api-version=2018-09-01
DEBUG: RequestCreated: https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/
backupLocations/local?api-version=2018-09-01
DEBUG: HeaderParametersAdded: https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup
.Admin/backupLocations/local?api-version=2018-09-01
```
At this time we do not know the actual host name, so we plan to hide it, and only keep the path beginning from `/subscriptions/...`

2. Second part is in the HTTP request
```log
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
GET

Absolute Uri:
https://management.azure.com/subscriptions/311d80f2-e4ed-492e-98ae-6961db2683f1/resourceGroups/system.local/providers/Microsoft.Backup.Admin/backupLocations/local?a
pi-version=2018-09-01
```

**Comment by isra-fel** [comment]
Root cause for 2. : the generated code to send HTTP request looks like this:

```csharp
await eventListener.Signal(Microsoft.Azure.PowerShell.Cmdlets.DedicatedHsm.Runtime.Events.BeforeCall, request);
_response = await sender.SendAsync(request, eventListener);
await eventListener.Signal(Microsoft.Azure.PowerShell.Cmdlets.DedicatedHsm.Runtime.Events.ResponseCreated, _response);
```

URI in `request` object is fake, so the [handler for `BeforeCall` event](https://github.com/Azure/azure-powershell/blob/91f1b7568ed7189ecb827f264c4265d053967442/src/Accounts/Accounts/CommonModule/AzModule.cs#L230) prints wrong log.

When the request is sent, in HTTP pipeline, it's URI is replaced to the real one according to which environment you are in, see https://github.com/Azure/azure-powershell/blob/91f1b7568ed7189ecb827f264c4265d053967442/src/Accounts/Accounts/CommonModule/ContextAdapter.cs#L158

**Comment by isra-fel** [comment]
Solutions:

1. ✔️  send `BeforeCall` event after the request is sent, in this way the URI in request is already replaced
    - But the event name would be somehow misleading
1. replace request URI in `BeforeCall`'s handler
    - Getting the real URI requires context info, which is not available there
1. do not log host name at all
    - But we still need it to debug

So it looks to me the first method is best.

**Comment by isra-fel** [comment]
So I chose the first solution and here's a sample output:

```powershell
PS C:\Users\yeliu\code\azure-powershell\src\DedicatedHsm [Az.DedicatedHsm]> Get-AzDedicatedHsm
DEBUG: Sought all Az modules and got latest version 5.2.0
DEBUG: 11:53:22 AM - GetAzureRMContextCommand begin processing with ParameterSet 'GetSingleContext'.
DEBUG: 11:53:22 AM - using account id 'yeliu@microsoft.com'...
DEBUG: AzureQoSEvent: CommandName - Get-AzContext; IsSuccess - True; Duration - 00:00:00.0603957;
DEBUG: Finish sending metric.
DEBUG: 11:53:23 AM - GetAzureRMContextCommand end processing.
DEBUG: CmdletBeginProcessing: 
DEBUG: CmdletProcessRecordStart: 
DEBUG: [CmdletProcessRecordAsyncStart]: Created new QosEvent for command 'Get-AzDedicatedHsm_List1': AzureQoSEvent: CommandName - Get-AzDedicatedHsm_List1; IsSuccess - False; Duration - 00:00:00;
DEBUG: CmdletProcessRecordAsyncStart:
DEBUG: CmdletGetPipeline:
DEBUG: CmdletBeforeAPICall: 
DEBUG: URLCreated: /subscriptions/9e223dbe-3399-4e19-88eb-0975f02ac87f/providers/Microsoft.HardwareSecurityModules/dedicatedHSMs?api-version=2018-10-31
DEBUG: RequestCreated: /subscriptions/9e223dbe-3399-4e19-88eb-0975f02ac87f/providers/Microsoft.HardwareSecurityModules/dedicatedHSMs?api-version=2018-10-31
DEBUG: HeaderParametersAdded:
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
GET

Absolute Uri:
https://management.autorest.com/subscriptions/9e223dbe-3399-4e19-88eb-0975f02ac87f/providers/Microsoft.HardwareSecurityModules/dedicatedHSMs?api-version=2018-10-31

Headers:
x-ms-unique-id                : 1
x-ms-client-request-id        : 40ec9fea-711a-45fd-90ce-cb5bc5cad59d
CommandName                   : Get-AzDedicatedHsm
FullCommandName               : Get-AzDedicatedHsm_List1
ParameterSetName              : __AllParameterSets
User-Agent                    : AzurePowershell/Az4.0.0-preview

Body:



DEBUG: BeforeCall: 
DEBUG: [ADAL]: Information: 2020-12-15T03:53:23.9773094Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: ADAL PCL.CoreCLR with assembly version '3.19.2.6005', file version '3.19.50302.0130' and informational version '2a8bec6c4c76d0c1ef819b55bdc3cda2d2605056' is running...
DEBUG: [ADAL]: Information: 2020-12-15T03:53:23.9782596Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: ADAL PCL.CoreCLR with assembly version '3.19.2.6005', file version '3.19.50302.0130' and informational version '2a8bec6c4c76d0c1ef819b55bdc3cda2d2605056' is running...
DEBUG: [ADAL]: Information: 2020-12-15T03:53:23.9788878Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: === Token Acquisition started:
        CacheType: null
        Authentication Target: User
        , Authority Host: login.microsoftonline.com
DEBUG: [ADAL]: Information: 2020-12-15T03:53:23.9790087Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: === Token Acquisition started:
        Authority: https://login.microsoftonline.com/72f988bf-86f1-41af-91ab-2d7cd011db47/
        Resource: https://management.core.windows.net/
        ClientId: 1950a258-227b-4e31-a9cf-717495945fc2
        CacheType: null
        Authentication Target: User

DEBUG: [ADAL]: Verbose: 2020-12-15T03:53:24.5037034Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: Loading from cache.DEBUG: [ADAL]: Verbose: 2020-12-15T03:53:24.5037743Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: Loading from cache.DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5144421Z: 00000000-0000-0000-0000-000000000000 - LoggerBase.cs: Deserialized 1 
items to token cache.
DEBUG: [ADAL]: Verbose: 2020-12-15T03:53:24.5161587Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: Looking up cache for a token...
DEBUG: [ADAL]: Verbose: 2020-12-15T03:53:24.5161830Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: Looking up cache for a token...
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5187228Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: An item matching the requested resource was found in the cache
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5187452Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: An item matching the requested resource was found in the cache
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5198810Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: 21.946969708333334 minutes left until token in cache expires
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5199017Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: 21.946969708333334 minutes left until token in cache expires
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5199310Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: A matching item (access token or refresh token or both) was found in the cache
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5199446Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: A matching item (access token or refresh token or both) was found in the cache
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5211542Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: === Token Acquisition finished successfully. An access token was returned: Expiration Time: 12/15/2020 4:15:21 AM +00:00
DEBUG: [ADAL]: Information: 2020-12-15T03:53:24.5211792Z: 9267fbc7-2434-44a1-9a0a-cd9434180b09 - LoggerBase.cs: === Token Acquisition finished successfully. An access token was returned: Expiration Time: 12/15/2020 4:15:21 AM +00:00Access Token Hash: dOOr/C6QENIASG43d1xYyTtv3ltfZt1oWL0zaJqQIPE=
         User id: d7e17135-d5a7-4b8b-89e5-252aa15b7e01
DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
NotFound

Headers:
Date                          : Tue, 15 Dec 2020 03:53:30 GMT
Connection                    : keep-alive
Keep-Alive                    : timeout=30
Server                        : Apache/2
X-Powered-By                  : PHP/7.4.10
Cache-Control                 : no-cache
X-DS-Version                  : 1530208438

Body:
Not Found


DEBUG: ResponseCreated: 
DEBUG: BeforeResponseDispatch:
DEBUG: Finally:
......
```

**Comment by isra-fel** [comment]
@bganapa Hi Bala, this issue has been fixed in the latest [autorest/powershell v3.0.409](https://www.npmjs.com/package/@autorest/powershell)

Please use `--use:@autorest/powershell@3.0.409` when generating.
